File: go/modules.go
package agentdisclosure

import "sort"

// --- Redaction (port of src/redaction.ts verifyRedacted) ------------------------
//
// A redacted view reveals only a subset of fields; the signature is over the
// per-field COMMITMENTS, not the cleartext, so withheld fields stay opaque but
// verifiable. The view is held as generic decoded-JSON maps so the canonical bytes
// match the TS reference exactly (meta + commitments under the signature, and each
// revealed value inside its commitment recompute).

// RedactedView is a holder view revealing only a subset of committed fields. meta,
// commitments, and the revealed values are kept as raw decoded-JSON for byte-exact
// canonicalization. Decode the fixture with json.Decoder.UseNumber().
type RedactedView struct {
	Meta        map[string]any
	Commitments map[string]any
	Revealed    map[string]any // field -> {value, salt}
	Signature   Signature
}

// commit reproduces the redaction commitment: sha256Hex(canonicalize(value)+":"+salt).
func commit(value any, salt string) string {
	return Sha256Hex(Canonicalize(value) + ":" + salt)
}

// VerifyRedacted verifies a redacted view (SPEC.md selective-disclosure):
//  1. meta.agentId equals signature.publicKey (identity binding),
//  2. the signature covers canonicalize({meta, commitments}),
//  3. each revealed field recomputes to its committed value.
//
// Returns (ok, revealedFields) where revealedFields is the sorted set of fields
// whose disclosure is cryptographically proven (empty when ok is false).
func VerifyRedacted(view RedactedView) (bool, []string) {
	agentID, _ := view.Meta["agentId"].(string)
	if agentID != view.Signature.PublicKey {
		return false, nil
	}

	signed := Canonicalize(map[string]any{"meta": view.Meta, "commitments": view.Commitments})
	if !VerifyMessage(signed, view.Signature.PublicKey, view.Signature.Value) {
		return false, nil
	}

	revealedFields := []string{}
	for field, raw := range view.Revealed {
		rv, _ := raw.(map[string]any)
		salt, _ := rv["salt"].(string)
		expected, ok := view.Commitments[field].(string)
		if !ok {
			return false, nil
		}
		if commit(rv["value"], salt) != expected {
			return false, nil
		}
		revealedFields = append(revealedFields, field)
	}
	sort.Strings(revealedFields)
	return true, revealedFields
}

// --- Revocation (port of src/revocation.ts verifyRevocation) --------------------

// SignedRevocation is a revocation attributable to the issuing key. The signed
// bytes are canonicalize({id, reason, revokedAt}).
type SignedRevocation struct {
	ID        string `json:"id"`
	Reason    string `json:"reason"`
	RevokedAt string `json:"revokedAt"`
	PublicKey string `json:"publicKey"`
	Signature string `json:"signature"`
}

// VerifyRevocation verifies a signed revocation against its embedded public key.
func VerifyRevocation(rec SignedRevocation) bool {
	body := Canonicalize(map[string]any{
		"id":        rec.ID,
		"reason":    rec.Reason,
		"revokedAt": rec.RevokedAt,
	})
	return VerifyMessage(body, rec.PublicKey, rec.Signature)
}

// --- Transparency (port of src/transparencyTransport.ts verifyInclusionProof) ---

// TransparencyLogEntry is a hash-linked transparency-log entry. Fields are kept as
// raw decoded-JSON (index as json.Number, prevHash as string-or-null) so the entry
// hash recomputes byte-exact. Decode the fixture with json.Decoder.UseNumber().
type TransparencyLogEntry struct {
	Index            any    // json.Number after UseNumber decode
	DisclosureDigest string
	AgentID          string
	IssuedAt         string
	PrevHash         any // string, or nil for the genesis entry
	Hash             string
}

// VerifyInclusionProof recomputes the entry hash from its own fields and confirms
// it matches the stored hash. This catches a log handing back a tampered entry; it
// does not prove the entry sits in the live chain.
func VerifyInclusionProof(entry TransparencyLogEntry) bool {
	expected := Sha256Hex(Canonicalize(map[string]any{
		"index":            entry.Index,
		"disclosureDigest": entry.DisclosureDigest,
		"agentId":          entry.AgentID,
		"issuedAt":         entry.IssuedAt,
		"prevHash":         entry.PrevHash,
	}))
	return expected == entry.Hash
}

File: python/agent_disclosure/modules.py
"""Ports of the redaction, revocation, and transparency-transport verifiers.

  - `verify_redacted`  -> `src/redaction.ts::verifyRedacted`
  - `verify_revocation`-> `src/revocation.ts::verifyRevocation`
  - `verify_inclusion_proof` -> `src/transparencyTransport.ts::verifyInclusionProof`

All three reuse the byte-matched `canonicalize` / `sha256_hex` and the asymmetric
`verify_message`, so they accept exactly the vectors the TS reference emits.
"""

from .canonical import canonicalize, sha256_hex
from .attestation import verify_message_hex


def _commit(value, salt: str) -> str:
    """Salted commitment for a field: sha256(canonicalize(value) + ':' + salt)."""
    return sha256_hex(f"{canonicalize(value)}:{salt}")


def verify_redacted(view: dict):
    """Verify a redacted view. Returns (ok, revealed_fields):
      1. meta.agentId == signature.publicKey (identity binding),
      2. signature covers canonicalize({meta, commitments}),
      3. each revealed field recomputes to its committed value.
    On any failure returns (False, []); on success revealed_fields is sorted."""
    meta = view["meta"]
    commitments = view["commitments"]
    signature = view["signature"]

    if meta["agentId"] != signature["publicKey"]:
        return False, []

    signed = canonicalize({"meta": meta, "commitments": commitments})
    if not verify_message_hex(signed, signature["publicKey"], signature["value"]):
        return False, []

    revealed_fields = []
    for field, payload in view["revealed"].items():
        expected = commitments.get(field)
        if expected is None:
            return False, []
        if _commit(payload["value"], payload["salt"]) != expected:
            return False, []
        revealed_fields.append(field)

    return True, sorted(revealed_fields)


def verify_revocation(record: dict) -> bool:
    """Verify a signed revocation against its embedded public key: ed25519 over
    canonicalize({id, reason, revokedAt})."""
    return verify_message_hex(
        canonicalize(
            {"id": record["id"], "reason": record["reason"], "revokedAt": record["revokedAt"]}
        ),
        record["publicKey"],
        record["signature"],
    )


def verify_inclusion_proof(entry: dict) -> bool:
    """Recompute the transparency-log entry hash from its own fields and confirm
    it matches: hash == sha256(canonicalize({index, disclosureDigest, agentId,
    issuedAt, prevHash}))."""
    expected = sha256_hex(
        canonicalize(
            {
                "index": entry["index"],
                "disclosureDigest": entry["disclosureDigest"],
                "agentId": entry["agentId"],
                "issuedAt": entry["issuedAt"],
                "prevHash": entry["prevHash"],
            }
        )
    )
    return expected == entry["hash"]
